finat/fiat_elements.py

```diff
@@ -173,6 +176,120 @@ def point_evaluation(self, order, refcoords, entity=None):
+        # order form a vector required for correct contraction with Q. Will
+        # become a FInAT.PointSet later.
+        # pt_hash = numpy.round(x_k, x_hash_decimals) such that
+        # numpy.allclose(x_k, pt_hash, atol=1*10**-dec) == true
+        x = {}
+        x_hash_decimals = 12
+        x_hash_atol = 1e-12  # = 1*10**-dec
+
+        #
+        # BUILD Q TENSOR
+        #
+
+        # FIXME: The below loop is REALLY SLOW for BDM - Q and x should just be output as the dual basis
+
+        last_shape = None
+        self.Q_is_identity = True  # TODO do this better
+
+        fiat_dual_basis = self._element.dual_basis()
+        # i are rows of Q
+        for i, dual in enumerate(fiat_dual_basis):
+            point_dict = dual.get_point_dict()
+            for j, (x_j, tuples) in enumerate(point_dict.items()):
+                x_j = PointSet((x_j,))  # TODO doesn't need to be PointSet
+                # get weights q_j for point x_j from complicated FIAT dual data
+                # structure
+                weight_dict = {cmp: weight for weight, cmp in tuples}
+                if len(self._element.value_shape()) == 0:
+                    q_j = np.array(weight_dict[tuple()])
+                else:
+                    q_j = np.zeros(self._element.value_shape())
+                    for key, item in weight_dict.items():
+                        q_j[key] = item
+
+                # Ensure all weights have the same shape
+                if last_shape is not None:
+                    assert q_j.shape == last_shape
+                last_shape = q_j.shape
+
+                # Create hash into x
+                x_hash = np.round(x_j.points, x_hash_decimals)
+                assert np.allclose(x_j.points, x_hash, atol=x_hash_atol)
+                x_hash = hash(tuple(x_hash.flatten()))
+
+                # Get value and index k or add to dict. k are the columns of Q.
+                try:
+                    x_j, k = x[x_hash]
+                except KeyError:
+                    k = len(x)
+                    x[x_hash] = x_j, k
+
+                # q_j may be tensor valued
+                it = np.nditer(q_j, flags=['multi_index'])
+                for q_j_entry in it:
+                    Q[(i, k) + it.multi_index] = q_j_entry
+                    if len(set((i, k) + it.multi_index)) > 1:
+                        # Identity has i == k == it.multi_index[0] == ...
+                        # Since i increases from 0 in increments of 1 we know
+                        # that if this check is not triggered we definitely
+                        # have an identity tensor.
+                        self.Q_is_identity = False
+
+        #
+        # CONVERT Q TO gem.Literal (TODO: should be a sparse tensor)
+        #
+
+        # temporary until sparse literals are implemented in GEM which will
+        # automatically convert a dictionary of keys internally.
+        Q = gem.Literal(sparse.as_coo(Q).todense())
```

Comment:
This is a regression which we should do some timing on. Things to test: 3D high order non-tensor product (such as a degree 10 tetrahedron)

Reply:
RT or BDM are good things to try

Reply:
(test against master! we've introduced a bunch of multiplying by zeros)

Reply:
looks to be fine - see for example https://github.com/firedrakeproject/firedrake/pull/2115#issuecomment-888394264